### PR TITLE
Better unicode to short performance

### LINKED ISSUE #45: emojione.mapShortToUnicode() is too computively expensive, it needs Memoization or precalculation
In the javascript version

*emojione.mapShortToUnicode()* is slow and creates tons useless of instances at each run. This is actually executed at every single call to public conversion api from utf8 to image (= very often in a single page app or a big page).
It would benefits greatly from memoization, the only drawback is the current api to add manual emoticon does not provide easy hook to invalid this memoization (no setters).

I had to monkeypatch my version of emojione to preserve decent execution timing: 

    emojione._mapShortToUnicode = emojione.mapShortToUnicode;
    emojione.mapShortToUnicode = function() {
      if (!emojione.memMapShortToUnicode) {
        emojione.memMapShortToUnicode = emojione._mapShortToUnicode();
      }
      return emojione.memMapShortToUnicode;
    };

PS: Regexp could also be precalculated instead of being generated every single execution but gain are much less visible in normal situation.

### PR DESCRIPTION
**Pull request by Coquevas**
Hi there, I just fixed #45 with this pull request.

All tests passing, after two fixes:

1. Cache burst param was not in the expectations for the URL test
2. As I said in #51, the `toShort` test suite fails due to the new flags. I did a workaround to test my code but didn't pushed it.

Performance enhancement can be checked in this JSPerf: http://jsperf.com/emojione-replacement-speed-test/3

The previous algorithm was O(c), where c is constant but large c: Emoji's list size (~1K). The new algorithm is O(n), but with n: Number of emojis in the string. So, worst case scenario is replacing a string with more than 1K emojis.

PS: I don't know how to generate the minified JS file, sorry.

### CODE REVIEW

File: lib/js/emojione.js [hunk 1]
```diff
@@ -425,13 +448,13 @@
...
 
-    ns.replaceAll = function(string, find, replaceWith) {
+    ns.replaceAll = function(string, find, replacementList) {
 
         var search = new RegExp("<object[^>]*>.*?<\/object>|<span[^>]*>.*?<\/span>|<(?:object|embed|svg|img|div|span|p|a)[^>]*>|("+find+")", "gi");
 
         // callback prevents replacing anything inside of these common html tags as well as between an <object></object> tag
         var replace = function(entire, m1) {
-            return ((m1 == undefined) || (m1 == '')) ? entire : replaceWith;
+            return replacementList[m1];
```
Comment (Coquevas) [inline lib/js/emojione.js]: Maybe I was too optimistic not checking for emptiness. Now I'm thinking to replace it for something like:

    return replacementList[m1] || m1;

**Comment by tomByrer** [comment]
I ran your tests on several browsers.  IE11 results seemed... strange.

**Comment by Coquevas** [comment]
Thanks @tomByrer, I'll run it on a VM with IE11 and try to fix it.

**Comment by maximeg** [comment]
Hi there :smile: 

Is there an ETA for landing this improvement in performance ? I may have several leads too if you want.

**Comment by mikebe11** [comment]
Hi @maximeg ,
There's no ETA on this. Time is somewhat limited. No telling when we can devote time to it. Not sure about the IE11 issue mentioned in a previous comment either.
What do mean by leads?

**Comment by Coquevas** [comment]
Hi @maximeg,

I'm having a busy year but I still want to resolve the issue and try to get this merged.

**Comment by mikebe11** [comment]
Hi @Coquevas 

Was looking to close this PR since these files can no longer be merged as they are. I don't want to abandon your idea for memoization though. I added your new version of toShort() to the latest release last night to check it out. It worked great. The only change I had to make was in now having to escape the find in `ns.replaceAll` using `ns.escapeRegExp`, but exclude the pipe so as not to ruin the or operator in the regex. Currently, this escaping is only needed for the keycap asterisk emoji. Then I ran it through benchmark.js to get the following.

screen shot 2016-02-04 at 4 19 15 pm

I still want to run some more tests, not for speed but coverage, and see how things look. I'll have to run some on IE as well to see what's going on there.

Anyway, do you think you may come back to this soon? In the meantime I can share what test results I get. Perhaps they'll be helpful.

**Comment by Coquevas** [comment]
Hi @mikebe11, that's great. I had to admit that the idea of this PR being separated from the main branch for so long didn't help me on finding time to fix it but now that you did the effort I'll put an eye on it today.

**Comment by Coquevas** [comment]
Hi again @mikebe11,

I rebased my changes on top of master, fixed all the conflicts, and, as you said, removed the OR operand from `ns.escapeRegExp` to make it work. Now is passing all the tests, but not tried on IE yet. Travis CI failed but looks like an environment issue.

**Comment by mikebe11** [comment]
I haven't gotten to my tests yet, IE or coverage. I'm currently having grunt issues in the Travis build, which was fine days ago. That's why it's currently failing. I'll pull your changes and check them out soon as I can.
Hopefully I can cherry pick your .js commit. The commit and file change counts have gotten overwhelming and I'm not comfortable enough with git to do a full merge here.

**Comment by mikebe11** [comment]
Hi @Coquevas ,
So I made a simple table to check the coverage of all the unicode characters. Good news is they all converted. Unfortunately, what happened is the shortname aliases overwrote the primary shortname to produce what's in the screencap. I'm pretty sure this was in `ns.mapEmojioneList`.
(screencap) primary shortname from emoji_strategy.json - unicode character - new ToShort() results

screen shot 2016-02-08 at 10 25 32 pm

Looks like a check for the existence of the unicode value is needed before calling `addToMapStorage()`.

**Comment by mikebe11** [comment]
Ran the library tests and also checked the coverage in IE 11. Tests passed and everything converted.

**Comment by Coquevas** [comment]
@mikebe11 so then the only problem is the flags one. I'll try to take a look today.

**Comment by mikebe11** [comment]
@Coquevas   Not just the flags. It's any that has a shortname alias. I don't know the exact count but emoji with aliases are in most, if not all, categories.
eg.  `:dove:':["1f54a"],':dove_of_peace:':["1f54a"]`

**Comment by Coquevas** [comment]
OK, I understand you now that I looked to the code.The problem here is that we need to choose a way to know if a shortname is an alias or is canonical.

**Comment by mikebe11** [comment]
The shortnames should be ordered in `ns.emojioneList` first by primary, then by any aliases that exist. My first thought is a check to see if an entry exists before `ns.memMapShortToUnicode[unicode] = shortname;` Although with the asynchronous nature of javascript not sure if that's a sufficient approach.

**Comment by Coquevas** [comment]
so, in your example, `:dove:':["1f54a"],':dove_of_peace:':["1f54a"]`, the canonical shortname is `:dove:` and the alias is `:dove_of_peace:`, and for US flags, `:flag_us:':["1f1fa-1f1f8"],':us:':["1f1fa-1f1f8"]` the canonical shortname is  `:flag_us:` and the alias is `:us:`, am I right?

**Comment by mikebe11** [comment]
Yes, that's correct. emoji.json and emoji_strategy.json are good references to see which is which.
eg.
`"flag_gr":{"unicode":"1f1ec-1f1f7","shortname":":flag_gr:","aliases":":gr:","keywords":"greece country nation ellas ellada gr flag other"}`

**Comment by Coquevas** [comment]
According to [ECMA-262 5.1 specification](http://www.ecma-international.org/ecma-262/5.1/#sec-12.6.4), quoted [here](http://stackoverflow.com/questions/280713/elements-order-in-a-for-in-loop), 

> The mechanics and order of enumerating the properties ... is not specified.

This was changed on [ECMA-262 6.0 specification](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys)

> 1. Each own property key that is an integer index, in ascending numeric index order
> 2. Each own property key that is a String but is not an integer index, in property creation order
> 3. Each own property key that is a Symbol, in property creation order

That means that you shouldn't rely on ordered iteration of an Object unless you are using a really new (2015) implementation of Javascript. Reality is that the only major browser that doesn't return String properties on creation order was Chrome, and [new versions are implementing 6.0 spec](https://bugs.chromium.org/p/v8/issues/detail?id=164#c140).

**tl;dr** - @mikebe11, are you OK assuming ordered iteration, with the small chance of returning aliases instead of canonical shortames in old browsers? If so, I can skip shortnames already added to `ns.memMapShortToUnicode`. Regarding your concerns, JavaScript is essentially single-threaded, you shouldn't worry about a different thread interrupting your method.

**Comment by mikebe11** [comment]
Probably not ok on assuming ordered iteration. Even though the chance is small I can see an issue eventually being submitted when the expected result doesn't show up for someone.
Gonna have to think on this one a bit.

**Comment by Coquevas** [comment]
Options that come to my mind right now:

**New `ns.canonicalEmojioneList` object similar to `ns.EmojioneList` but without aliases**
- Pros:
  - Pre-generated
  - Small logic changes, only iterating a different object where needed
- Cons:
  - One more big list on the file. Library size increased.

**Replace `ns.EmojioneList` object with an array (order guaranteed) of objects**

````javascript
ns.emojioneList = [
    {":bikini:": ["1f459"]},
    {":womans_clothes:": ["1f45a"]}
]
````
- Pros:
  - Ordered `¯\_(ツ)_/¯`
- Cons:
  - More changes to the current logic, even in non-related methods.
  - No O(1) access to unicode anymore

**Replace `ns.EmojioneList` plain text values with objects**

````javascript
ns.emojioneList = {
    ":dove:": {
        "shortname": ":dove:",
        "unicode": ["1f54a"],
        "isCanonical": true
    },
    ":dove_of_peace:": {
        "shortname": ":dove_of_peace:",
        "unicode": ["1f54a"],
        "isCanonical": false
    }
}
`````

- Pros:
  - Kind of compromise solution.
- Cons:
  - More changes to the current logic, but not as heavy as previous one.

**Comment by mikebe11** [comment]
Without committing to anything, I'm leaning towards  **Replace** `ns.EmojioneList` **plain text values with objects**.  Revisions of non-related methods shouldn't be too bad.

**Comment by Coquevas** [comment]
I can try to do it and see how it looks. Is that list generated somehow? If not, that could be a good improvement. Single source of true and code generation for every supported language.

**Comment by mikebe11** [comment]
@Coquevas  Here are a couple of new ns.EmojioneList variables if you want to try them out. We have a script that generates the list from our database which I altered to get these. I didn't test them, just ran them through jsonlint.

I included a version without the shortname property. As I was altering the script I realized it was redundant since the key is the shortname. Saves about 55,000 characters.

Another approach is if you kept the shortname property it could be the canonical shortname, then the "isCanonical" property isn't really needed. An equality check between the key and shortname property would determine whether it's the canonical shortname.

[Coquevas.zip](https://github.com/Ranks/emojione/files/133748/Coquevas.zip)

**Comment by mikebe11** [comment]
@Coquevas  Just checking to see if you've tried the new ns.emojioneList variables yet?

**Comment by Coquevas** [comment]
@mikebe11 sorry for the late reply. I tested your new `ns.emojioneList` variables and I chose the version without the `shortname` property. Everything is working but two small issues. 

The first one was that the `unicode` property was an Array for shortnames with more than one code point but was a string for the shortnames with only one code point. I did the changes by myself to standardise it to be always an Array.

The second one is that I have a merge conflict at `ns.emojioneList` variable with your master branch, probably because you added more shortnames, so I'd need you to generate a new list that I can replace and fix the conflict.

Thanks!

**Comment by mikebe11** [comment]
@Coquevas No problem. Glad you’re still alive and kicking.
My bad on the property type, array/string. I’ll change my script to always have an array there. I’ll get that updated data to you very soon. Thanks for this. Looking forward to getting your changes merged. It’ll be a nice improvement.

**Comment by Coquevas** [comment]
@mikebe11 a nice and long running one ;)

**Comment by mikebe11** [comment]
Here's the updated `ns.emojioneList`. Along with this variable I expect there will be other conflicts like the new classes from #203 which was recently merged.
[emojioneList.zip](https://github.com/Ranks/emojione/files/185240/emojioneList.zip)

**Comment by Coquevas** [comment]
@mikebe11 done! \o/

**Comment by mikebe11** [comment]
Sweet! Will check it out asap.

**Comment by mikebe11** [comment]
@Coquevas Thanks very much for the time spent on this. I went through it and ran tests. Everything was good. I hope I didn't overlook anything.

I didn't end up merging from here. There were so many additional file changes in the last commit to bring things in line with the master branch. Probably would've been ok but I'm not that experienced with git so I was concerned about unintended changes. I ended up cherry picking the latest commit related to the js file. 

Great work on this. Cheers

**Comment by Coquevas** [comment]
:+1: :ok_hand: 

@mikebe11 I also added a test case to `tests.html` on https://github.com/Ranks/emojione/pull/53/commits/6846954a1b15fdd6160416c7d4086db4804e9256 to make sure the canonical name is used, in case you want to cherry pick it too.

**Comment by mikebe11** [comment]
I did notice that test in 07806469286414d982cb244a873f5c71c1a68868. Since it didn't get included when I cherry picked that commit I added it in my release update commit. I added your test to `validate.json` and regenerated `tests.html` with the included script, `generate.js`.

**Comment by Coquevas** [comment]
Then it's done! Thank you :)

**Comment by mikebe11** [comment]
@Coquevas I was looking into another issue and found a bug with the new toShort implementation. Converting `<a href="emojione.com">😃</a> more text` returns `undefined:smiley: more text` so the tags that should be excluded on line 498 are being returned as undefined.

I came up with the following that appears to fix it, a revision of the original line. Replace line 503 with `return ((typeof m1 === 'undefined') || (m1 === '')) ? entire : replacementList[m1];`

Would you take a look at this and give me your thoughts?

Thanks

**Comment by Coquevas** [comment]
@mikebe11 good catch.

What about this?
`return m1 ? replacementList[m1] : entire;`

I didn't try it, but is a small refactor of yours including any _falsy_ value (Null, undefined, empty string, NaN...)

**Comment by mikebe11** [comment]
@Coquevas It probably covers more falsy values than is necessary but it does work and the extra isn't hurting anything. I'll push your refactor to go along with your recent update. Thanks for the input.

**Comment by Coquevas** [comment]
@mikebe11 you can call it Premature Optimisation, I call it Readable Code :wink:
